haystack/nodes/prompt/prompt_node.py

```diff
@@ -681,11 +217,28 @@ def __init__(
 
         self.model_kwargs = model_kwargs if model_kwargs else {}
 
-        self.invocation_layers: List[Type[PromptModelInvocationLayer]] = []
+        self.invocation_layer_classes: List[Type[PromptModelInvocationLayer]] = known_providers()
+        if invocation_layer_class:
+            klass: Optional[Type[PromptModelInvocationLayer]] = None
+            if isinstance(invocation_layer_class, str):
+                # try to find the invocation_layer_class provider class
+                search_path: List[str] = [
+                    f"haystack.nodes.prompt.providers.{invocation_layer_class}",
+                    invocation_layer_class,
+                ]
+                klass = next((pydoc.locate(path) for path in search_path if pydoc.locate(path)), None)  # type: ignore
+
+                if not klass:
+                    raise ValueError(
+                        f"Could not locate PromptModelInvocationLayer class with name {invocation_layer_class}. "
+                        f"Make sure to pass the full path to the class."
+                    )
+
+            if not issubclass(klass, PromptModelInvocationLayer):
+                raise ValueError(f"Class {invocation_layer_class} is not a subclass of PromptModelInvocationLayer.")
 
-        self.register(HFLocalInvocationLayer)  # pylint: disable=W0108
-        self.register(OpenAIInvocationLayer)  # pylint: disable=W0108
-        self.register(AzureOpenAIInvocationLayer)  # pylint: disable=W0108
+            logger.info("Registering custom invocation layer class %s", klass)
+            self.register(klass)
```

Comment:
If I understand correctly, this is the part of the code that is registering additional `PromptModelInvocationLayers` in case the user provides one. 

If this is the case, there is a way to "automatically" register invocation layers: we should use `__init_subclass__`, which is the way `BaseComponent` currently uses to allow users to define custom nodes. The implementation is not straightforward, but it boils down to the following steps:
1. `PromptModelInvocationLayer` defines a method called `__init_subclass__`. This is called when a subclass is *imported*. This method will only register the class into a class variable of `PromptModelInvocationLayer`:
```python
class PromptModelInvocationLayer:
   ...
    def __init_subclass__(cls**kwargs):
        super().__init_subclass__(**kwargs)
        cls.providers[cls.__name__] = cls
    ....
```
2. `PromptModelInvocationLayer` then defines a classmethod called `get_provider()` which looks up into the table we built before:
```python
class PromptModelInvocationLayer:
   ...
    @classmethod
    def get_provider(cls, name):
        return cls.providers[name]
    ....
```
3. Once these two methods are set up, this entire block I highlighted can be removed. Every `PromptModelInvocationLayer` will automatically have access to a class attribute containing a mapping of all the subclasses of `PromptModelInvocationLayer` that you can access to load the provider you need. 

You won't even need the `invocation_layer_class` parameter: the class is going to be already registered in the providers map. However we can still keep it in order to allow users to force a provider instead of another. In that case we will need only the class name, not the full import path.

It's not the most straightforward Python code ever but it's very neat imho. Let me know if I managed to explain it and if you need help implementing this solution.